apps/protocol-frontend/src/hooks/useContributionBulkMint.ts

```diff
@@ -0,0 +1,114 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import useGovrnToast from '../components/toast';
+import { useNetwork, useSigner } from 'wagmi';
+import { useUser } from '../contexts/UserContext';
+import { networks } from '../utils/networks';
+import { ethers } from 'ethers';
+import { ChainIdError } from '@govrn/protocol-client';
+import pluralize from 'pluralize';
+
+type BulkMintOptions = {
+  ipfsContentUri: string;
+  engagementDate: Date | string;
+  attestations: { id: number }[];
+  guildName: string;
+  name: string;
+  action: string;
+  details?: string | null;
+  id: number;
+  proof?: string | null;
+  txHash?: string | null;
+  user: { id: number };
+  date_of_submission: Date | string;
+  activityTypeId: number;
+  status: { id: number; name: string };
+};
+
+const useContributionBulkMint = () => {
+  const toast = useGovrnToast();
+  const queryClient = useQueryClient();
+  const { data: signer } = useSigner();
+  const { chain } = useNetwork();
+  const { userData, govrnProtocol: govrn } = useUser();
+
+  const { mutateAsync, isLoading, isError, isSuccess } = useMutation(
+    ['bulk-mint-contribution'],
+    async (contributions: BulkMintOptions[]) => {
+      if (!(signer && chain?.id && userData)) {
+        throw new Error('What should the message be!?');
+      }
+
+      return await govrn.contribution.bulkMint(
+        {
+          address: networks[chain?.id].govrnContract,
+          chainId: chain?.id,
+          name: networks[chain?.id].name,
+        },
+        signer,
+        userData.address,
+        contributions.map(c => ({
+          id: c.id,
+          activityTypeId: c.activityTypeId,
+          userId: userData.id,
+          args: {
+            detailsUri: ethers.utils.toUtf8Bytes(c.ipfsContentUri),
+            dateOfSubmission: new Date(c.date_of_submission).getTime(),
+            dateOfEngagement: new Date(c.engagementDate).getTime(),
+          },
+          name: ethers.utils.toUtf8Bytes(c.name),
+          details: ethers.utils.toUtf8Bytes(c.details || ''),
+          proof: ethers.utils.toUtf8Bytes(c.proof || ''),
+        })),
+      );
+    },
+    {
+      onSuccess: result => {
+        /* Failed writes/updates to db are included in result. */
+
+        const minted = result.filter(i => i.status === 'fulfilled');
+
+        if (minted.length > 0) {
+          queryClient.invalidateQueries(['contributionList']);
+          queryClient.invalidateQueries(['contributionInfiniteList']);
+
+          toast.success({
+            title: 'Contribution Successfully Minted',
+            description: `${pluralize(
+              'Contribution',
+              minted.length,
+              true,
+            )} has been minted.`,
+          });
+        }
+
+        const failedToMint = result
+          .filter((i): i is PromiseRejectedResult => i.status === 'rejected')
```

Comment:
Why do we need to explicit about the type here and not line 68

Reply:
`filter` returns the same type as its input.
For successful mints, we only care about the count. 
for failed mints, we are interested in the `PromiseRejectedResult.reason`. Consequently, we need to narrow down the type that's why I added the type predicate `i is PromiseRejectedResult`


I found out that there is narrowing-down [`filter`](https://github.com/microsoft/TypeScript/blob/4f54e7e947298162d29f3104265e74dcfbc90d82/lib/lib.es5.d.ts#L1238-L1243) But it works with inheritance, Not sure if this applies to type union also or not.

Reply:
Both ways are not perfectly safe!
```ts
const failedToMint = result
  .filter<PromiseRejectedResult>((i): i is PromiseRejectedResult => {
                                   // ^ forces to put type predicate! (bonus)
    return i.status === 'rejected';
        // ^ You can change the predicate, but it won't catch that it's different type.
})
```

Reply:
Ah I, think the easiest way to narrow down the type is to use a for loop and rely on type inference. I am always wary with type hinting because it tends to cause problems eventually